packages/core/src/binary-data/file-system.manager.ts

```diff
@@ -165,10 +187,35 @@ export class FileSystemManager implements BinaryData.Manager {
 	 * The legacy ID format `{executionId}{uuid}` for `filesystem` mode is
 	 * no longer used on write, only when reading old stored execution data.
 	 */
-	private toFileId(workflowId: string, executionId: string) {
-		if (!executionId) executionId = 'temp'; // missing only in edge case, see PR #7244
+	private toFileId(location: BinaryData.FileLocation) {
+		return `${this.toRelativePath(location)}/binary_data/${uuid()}`;
+	}
+
+	private toRelativePath(location: BinaryData.FileLocation) {
+		switch (location.type) {
+			case 'execution': {
+				const executionId = location.executionId || 'temp'; // missing only in edge case, see PR #7244
+				return `workflows/${location.workflowId}/executions/${executionId}`;
+			}
+			case 'chat-hub-message-attachment':
+				return `chat-hub/sessions/${location.sessionId}/messages/${location.messageId}`;
+		}
+	}
```

Comment:
The binary data service is a storage layer, so it should have no knowledge know about app features. Can we implement this such that each feature defines and provides its own path format?

Reply:
My reasoning is mapping `FileLocation` to physical file path is within the responsibility of binary file manager implementations, because different implementation may require different format etc.

Reply:
Replied internally: https://n8nio.slack.com/archives/C0789EN39RC/p1763543953058649?thread_ts=1763543266.820279&cid=C0789EN39RC